kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1218 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Dynamic Process Loader for OTA application loads and updates
+//!
+//! These functions facilitate dynamic application loading and process
+//! creation during runtime without requiring the user to restart the
+//! device.
+
+use core::cell::Cell;
+
+use crate::capabilities::ProcessManagementCapability;
+use crate::config;
+use crate::create_capability;
+use crate::debug;
+use crate::hil::nonvolatile_storage::{NonvolatileStorage, NonvolatileStorageClient};
+use crate::kernel::Kernel;
+use crate::platform::chip::Chip;
+// use crate::process::ProcessId;
+use crate::process::{self, Process, ShortId};
+use crate::process_binary::{ProcessBinary, ProcessBinaryError};
+use crate::process_loading::ProcessLoadError;
+use crate::process_policies::ProcessFaultPolicy;
+use crate::process_standard::ProcessStandard;
+use crate::utilities::cells::{MapCell, OptionalCell, TakeCell};
+use crate::utilities::leasable_buffer::SubSliceMut;
+use crate::ErrorCode;
+
+const MAX_PROCS: usize = 10; //need this to store the start addresses of processes to write padding
+pub const BUF_LEN: usize = 512;
+const TBF_HEADER_LENGTH: usize = 16;
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum State {
+    Idle,
+    Setup,
+    AppWrite,
+    Load,
+    PaddingWrite,
+    Fail,
+}
+
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum PaddingRequirement {
+    PrePad,
+    PostPad,
+    PreAndPostPad,
+}
+
+/// This interface supports loading processes at runtime.
+pub trait DynamicProcessLoading {
+    /// Call to request loading a new process.
+    ///
+    /// This informs the kernel we want to load a process and the size of the entire process binary.
+    /// The kernel will try to find a suitable location in flash to store said process.
+    ///
+    /// Return value:
+    /// - `Ok((start_address, length))`: If there is a place to load the
+    ///   process, the function will return `Ok()` with the address to start at
+    ///   and the size of the region to store the process.
```

Comment:
This comment is out of date. I'm guessing it just returns the size?

Reply:
Now it returns size and if the process loader still has to do setup. Based on the value, the client app has to wait on the setup done callback.